### PR TITLE
[Clang][CodeGen] Optimised LLVM IR for atomic increments/decrements on floats

### LINKED ISSUE #53079: [clang] produces incorrect LLVM IR for atomic post increments on floats
https://godbolt.org/z/f8668T31M

```
auto test0()
{
    static _Atomic float n;
    return __c11_atomic_fetch_add(&n, 1, 5);
}

auto test1()
{
    static _Atomic float n;
    return n++;
}

```
test0 -  generates `atomicrmw fadd` and correct intel assembly (there could be a slight improvement here)
test1 - generates an LLVM IR code sequence using atomic compare exchange `cmpxchg` instead of using `atomicrmw fadd`. This code sequence has a bug in that it only returns the value of the first load from the atomic float. If two threads call `test1()` at the same time they could get the same value returned (`%2` below), rather than n and n + 1. The generated compare exchange LLVM IR could be fixed by returning `%11`, but now that `atomicrmw fadd` exists it should be used instead.

if `float` is changed to `int` both functions generate `atomicrmw add`

LLVM IR of `test1()`
```
  %1 = load atomic i32, i32* bitcast (float* @_ZZ5test1vE1n to i32*) seq_cst, align 4, !dbg !32, !tbaa !33
  %2 = bitcast i32 %1 to float, !dbg !32
  br label %3, !dbg !32

3: ; preds = %3, %0
  %4 = phi float [ %2, %0 ], [ %11, %3 ], !dbg !32
  %5 = fadd float %4, 1.000000e+00, !dbg !32
  %6 = bitcast float %4 to i32, !dbg !32
  %7 = bitcast float %5 to i32, !dbg !32
  %8 = cmpxchg i32* bitcast (float* @_ZZ5test1vE1n to i32*), i32 %6, i32 %7 seq_cst seq_cst, align 4, !dbg !32
  %9 = extractvalue { i32, i1 } %8, 0, !dbg !32
  %10 = extractvalue { i32, i1 } %8, 1, !dbg !32
  %11 = bitcast i32 %9 to float, !dbg !32
  br i1 %10, label %12, label %3, !dbg !32

12: ; preds = %3
  ret float %2, !dbg !36
```

should return `%11` not `%2`

### PR DESCRIPTION
**Pull request by Krishna-13-cyber**
Fixes #53079

### CODE REVIEW

File: clang/test/CodeGen/X86/x86-atomic-float.c [hunk 1]
```diff
@@ -0,0 +1,97 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 4
+// RUN: %clang_cc1 -triple x86_64-linux-gnu -target-cpu core2 %s -S -emit-llvm -o - | FileCheck %s
+// RUN: %clang_cc1 -triple i686-linux-gnu -target-cpu core2 %s -S -emit-llvm -o - | FileCheck -check-prefix=CHECK32 %s
```
Comment (RKSimon) [inline clang/test/CodeGen/X86/x86-atomic-float.c:3]: Reduce a lot of duplicate checks:
```
// RUN: %clang_cc1 -triple x86_64-linux-gnu -target-cpu core2 %s -S -emit-llvm -o - | FileCheck -check-prefixes=CHECK,CHECK64 %s
// RUN: %clang_cc1 -triple i686-linux-gnu -target-cpu core2 %s -S -emit-llvm -o - | FileCheck -check-prefixes=CHECK,CHECK32 %s
```

File: clang/test/CodeGen/X86/x86-atomic-float.c [hunk 2]
```diff
@@ -0,0 +1,64 @@
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 4
+// RUN: %clang_cc1 -triple x86_64-linux-gnu -target-cpu core2 %s -S -emit-llvm -o - | FileCheck %s
```
Comment (RKSimon) [inline clang/test/CodeGen/X86/x86-atomic-float.c:2]: Please can you add the i686 test coverage back? I mean that in most cases you should be able to share the check-prefixes

Comment (Krishna-13-cyber) [reply clang/test/CodeGen/X86/x86-atomic-float.c]: Yes I added the test, I misinterpreted the earlier review.
Thanks!

**Comment by Krishna-13-cyber** [comment]
Hi @RKSimon, any more changes required for this!?

File: clang/test/CodeGen/X86/x86-atomic-float.c [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+float test_float_post_inc()
+{
+    static _Atomic float n;
+    return n++;
+}
+
+// CHECK-LABEL: define dso_local float @test_float_post_dc(
+// CHECK-SAME: ) #[[ATTR0]] {
+// CHECK-NEXT:  entry:
+// CHECK-NEXT:    [[TMP0:%.*]] = atomicrmw fsub ptr @test_float_post_dc.n, float -1.000000e+00 seq_cst, align 4
```
Comment (RKSimon) [inline clang/test/CodeGen/X86/x86-atomic-float.c:33]: `fsub x, -1.0`? Should it be `fsub x, 1.0`?

**Review by RKSimon — APPROVED** [review]
LGTM

**Comment by Krishna-13-cyber** [comment]
> LGTM

Thanks!

**Pull request merged**